hoomd/md/pair/alch.py

```diff
@@ -0,0 +1,220 @@
+# Copyright (c) 2009-2021 The Regents of the University of Michigan
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause
+# License.
+
+"""Alchemical potentials."""
+
+from hoomd.logging import log, Loggable
+from hoomd.operation import _HOOMDBaseObject
+from hoomd.data.parameterdicts import TypeParameterDict
+from hoomd.data.typeconverter import SetOnce
+from hoomd.md.pair import pair
+
+# alternative access? hoomd.alch.md.pair would be nice to somehow link here
+# without making overly complicated
+# I just like alch(MD/MC) as alchemy specification aliases
+
+# TODO: remove this variable when the log names are finalized
+_api = ['value', 'mass', 'mu', 'avalue', 'amomentum', 'net_aforce', 'aforces']
+
+
+class _AlchemicalPairPotential(Loggable):
+
+    def __new__(cls, name, superclasses, attributedict):
+        attributedict[
+            '_cpp_class_name'] = 'Alchemical' + superclasses[0]._cpp_class_name
+        superclasses += (_AlchemicalMethods,)
+        return super().__new__(cls, name, superclasses, attributedict)
+
+    def __init__(cls, name, superclasses, attributedict):
+        cls._reserved_default_attrs['_alchemical_parameters'] = list
+        cls._accepted_modes = ('none', 'shift')
+        super().__init__(name, superclasses, attributedict)
+
+
+# Perhaps define a true base alch particle to base this off of just adding the
+# MD components mirroring the cpp side
+class AlchemicalMDParticle(_HOOMDBaseObject):
+    __slots__ = ['force', 'name', 'typepair', '_cpp_obj', '_mass']
+
+    def __new__(cls,
+                force: _AlchemicalPairPotential,
+                name: str = '',
+                typepair: tuple = None,
+                mass: float = 1.0):
+        typepair = tuple(sorted(typepair))
+        if (typepair, name) in force.alchemical_particles:
+            return force.alchemical_particles[typepair, name]
+        return super().__new__(cls)
```

Comment:
These objects are only defined relative to the pair potential they are in relation too. They're singletons in this regard. So here's a messy way to mask them that at least works. This doesn't allow delayed instantiation if the the force doesn't exist, but it doesn't have to be attached. Perhaps would be better to hide these methods?